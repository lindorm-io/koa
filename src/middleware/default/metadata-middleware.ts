import { KoaContext, Middleware } from "../../typing";
import { v1 as uuidv1, v4 as uuidv4 } from "uuid";

export const metadataMiddleware: Middleware<KoaContext> = async (ctx, next): Promise<void> => {
  ctx.metadata = {
    clientEnvironment: ctx.get("X-Client-Environment") || null,
    clientId: ctx.get("X-Client-ID") || null,
    clientName: ctx.get("X-Client-Name") || null,
    clientPlatform: ctx.get("X-Client-Platform") || null,
    clientVersion: ctx.get("X-Client-Version") || null,
    correlationId: ctx.get("X-Correlation-ID") || uuidv1(),
    deviceId: ctx.get("X-Device-ID") || null,
    installationId: ctx.get("X-Installation-ID") || null,
    sessionId: ctx.get("X-Session-ID") || uuidv4(),
  };

  ctx.metadataHeaders = {
    "X-Client-Environment": ctx.metadata.clientEnvironment,
    "X-Client-ID": ctx.metadata.clientId,
    "X-Client-Name": ctx.metadata.clientName,
    "X-Client-Platform": ctx.metadata.clientPlatform,
    "X-Client-Version": ctx.metadata.clientVersion,
    "X-Correlation-ID": ctx.metadata.correlationId,
    "X-Device-ID": ctx.metadata.deviceId,
    "X-Installation-ID": ctx.metadata.installationId,
    "X-Session-ID": ctx.metadata.sessionId,
  };

  await next();
};
