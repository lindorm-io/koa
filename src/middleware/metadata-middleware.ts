import { IKoaAppContext } from "../typing";
import { Middleware, Next } from "koa";
import { v1 as uuidv1, v4 as uuidv4 } from "uuid";

export const metadataMiddleware: Middleware = async (ctx: IKoaAppContext, next: Next): Promise<void> => {
  ctx.metadata = {
    clientEnvironment: ctx.get("X-Client-Environment") || null,
    clientId: ctx.get("X-Client-ID") || null,
    clientName: ctx.get("X-Client-Name") || null,
    clientPlatform: ctx.get("X-Client-Platform") || null,
    clientVersion: ctx.get("X-Client-Version") || null,
    correlationId: ctx.get("X-Correlation-ID") || uuidv1(),
    deviceId: ctx.get("X-Device-ID") || null,
    installationId: ctx.get("X-Installation-ID") || null,
    sessionId: ctx.get("X-Session-ID") || uuidv4(),
  };

  await next();
};
