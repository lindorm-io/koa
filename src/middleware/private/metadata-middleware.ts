import { KoaContext, Middleware } from "../../typing";
import { v1 as uuidv1, v4 as uuidv4 } from "uuid";
import { Environment } from "../../enum";

export const metadataMiddleware: Middleware<KoaContext> = async (ctx, next): Promise<void> => {
  ctx.metadata = {
    clientEnvironment: (ctx.get("X-Client-Environment") as Environment) || null,
    clientId: ctx.get("X-Client-ID") || null,
    clientPlatform: ctx.get("X-Client-Platform") || null,
    clientVersion: ctx.get("X-Client-Version") || null,
    correlationId: ctx.get("X-Correlation-ID") || uuidv1(),
    deviceId: ctx.get("X-Device-ID") || null,
    deviceUniqueId: ctx.get("X-Device-Unique-ID") || null,
    installationId: ctx.get("X-Installation-ID") || null,
    sessionId: ctx.get("X-Session-ID") || uuidv4(),
  };

  ctx.metadataHeaders = {
    "X-Client-Environment": ctx.metadata.clientEnvironment,
    "X-Client-ID": ctx.metadata.clientId,
    "X-Client-Platform": ctx.metadata.clientPlatform,
    "X-Client-Version": ctx.metadata.clientVersion,
    "X-Correlation-ID": ctx.metadata.correlationId,
    "X-Device-ID": ctx.metadata.deviceId,
    "X-Device-Unique-ID": ctx.metadata.deviceUniqueId,
    "X-Installation-ID": ctx.metadata.installationId,
    "X-Session-ID": ctx.metadata.sessionId,
  };

  await next();
};
