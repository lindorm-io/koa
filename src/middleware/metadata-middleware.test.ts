import { metadataMiddleware } from "./metadata-middleware";

jest.mock("uuid", () => ({
  v1: jest.fn(() => "0dacd3cc-035e-4968-881e-1fc30d496121"),
  v4: jest.fn(() => "a26dad28-e854-447d-bce6-5c685cddfea8"),
}));

describe("correlationIdMiddleware", () => {
  let ctx: any;
  let next: any;

  beforeEach(() => {
    ctx = {
      get: jest.fn((name) => name),
      userAgent: {
        browser: "browser",
        geoIp: { geoIp: 1 },
        os: "os",
        platform: "platform",
        source: "source",
        version: "version",
      },
    };
    next = () => Promise.resolve();
  });

  test("should use correlation id from header if it exists", async () => {
    await expect(metadataMiddleware(ctx, next)).resolves.toBe(undefined);
    expect(ctx.metadata).toMatchSnapshot();
  });

  test("should use uuid if header does not exist", async () => {
    ctx.get = jest.fn(() => undefined);

    await expect(metadataMiddleware(ctx, next)).resolves.toBe(undefined);
    expect(ctx.metadata).toMatchSnapshot();
  });

  test("should set user agent data", async () => {
    await expect(metadataMiddleware(ctx, next)).resolves.toBe(undefined);
    expect(ctx.agent).toMatchSnapshot();
  });

  test("should set null if agent data does not exist", async () => {
    ctx.userAgent = undefined;

    await expect(metadataMiddleware(ctx, next)).resolves.toBe(undefined);
    expect(ctx.agent).toMatchSnapshot();
  });
});
