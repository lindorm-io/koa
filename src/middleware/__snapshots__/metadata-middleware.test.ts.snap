// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`metadataMiddleware should use correlation id from header if it exists 1`] = `
Object {
  "clientEnvironment": "X-Client-Environment",
  "clientId": "X-Client-ID",
  "clientName": "X-Client-Name",
  "clientPlatform": "X-Client-Platform",
  "clientVersion": "X-Client-Version",
  "correlationId": "X-Correlation-ID",
  "deviceId": "X-Device-ID",
  "installationId": "X-Installation-ID",
  "sessionId": "X-Session-ID",
}
`;

exports[`metadataMiddleware should use uuid if header does not exist 1`] = `
Object {
  "clientEnvironment": null,
  "clientId": null,
  "clientName": null,
  "clientPlatform": null,
  "clientVersion": null,
  "correlationId": "0dacd3cc-035e-4968-881e-1fc30d496121",
  "deviceId": null,
  "installationId": null,
  "sessionId": "a26dad28-e854-447d-bce6-5c685cddfea8",
}
`;
